### ISSUE TITLE
Variable implicitly has type 'any' in some locations where its type cannot be determined in 5.0

### ISSUE THREAD

**Issue by jakebailey** [microsoft/TypeScript#52401]
# Bug Report

### 🔎 Search Terms

locations where its type cannot be determined

### 🕗 Version & Regression Information

- This changed between versions 4.9 and 5.0

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/PTAEFkE9QYwewCYFNQHM5IM6gBZIE5JA)

### 💻 Code

```ts repro
let foo;
try {
	if (Math.random() > 0.5) {
		foo = 1234;
	}
} catch {
	foo;
}
```

### 🙁 Actual behavior

Errors like:

```
Variable 'foo' implicitly has type 'any' in some locations where its type cannot be determined.(7034)
```

### 🙂 Expected behavior

No errors

**Comment by jakebailey** [comment]
Uh oh. Seems like I should revert that for now.

**Comment by jakebailey** [comment]
I should note that I don't actually know if the error is good or not, it's just new as of my PR above and that wasn't the intent of the change, so I've sent a PR to revert it for now and we can figure out the rest later.

**Comment by fatcerberus** [comment]
Isn't this a correct error?  `foo` has an evolving type and can theoretically be accessed (in this case, in the `catch` clause) without being initialized, which is what that error generally means.

**Comment by jakebailey** [comment]
Yes, I think that it is (and that's what I'm hearing from the team), but I definitely didn't intend to expose these errors in my catch clause PR (and definitely not inconsistently).
